mypyc/irbuild/prebuildvisitor.py

```diff
@@ -68,10 +71,29 @@ def __init__(
         # Map function to indices of decorators to remove
         self.decorators_to_remove: dict[FuncDef, list[int]] = decorators_to_remove
 
+        # A mapping of import groups (a series of Import nodes with
+        # nothing inbetween) where each group is keyed by its first
+        # import node.
+        self.module_import_groups: dict[Import, list[Import]] = {}
+        self._current_import_group: Import | None = None
+
         self.errors: Errors = errors
 
         self.current_file: MypyFile = current_file
 
+    def visit(self, o: Node) -> bool:
+        if isinstance(o, Import):
+            if self._current_import_group is not None:
+                self.module_import_groups[self._current_import_group].append(o)
+            else:
+                self.module_import_groups[o] = [o]
+                self._current_import_group = o
+            # Don't recurse into the import's assignments.
+            return False
```

Comment:
What happens in cases like this:
```
def f() -> None:
    import x
import y
```
Is it possible that the next import is part of an outer block?

Reply:
Ah, good catch! The old grouping logic would indeed group the two imports together. I've harden the implementation in https://github.com/python/mypy/pull/14917/commits/53d40b181ec61555b8d4a4f5c6a397441a32723b.